src/Moq.Analyzers/EnumerableExtensions.cs

```diff
@@ -0,0 +1,42 @@
+﻿namespace Moq.Analyzers;
+
+internal static class EnumerableExtensions
+{
+    /// <inheritdoc cref="SingleWhenOnly{TSource}(IEnumerable{TSource}, Func{TSource, bool})"/>
+    public static TSource? SingleWhenOnly<TSource>(this IEnumerable<TSource> source)
+    {
+        return source.SingleWhenOnly(_ => true);
+    }
+
+    /// <summary>
+    /// Returns the only element of a sequence that satisfies a specified condition,
+    /// and returns a default value if no such element exists or there are multiple matches.
+    /// </summary>
+    /// <typeparam name="TSource">The type of the <paramref name="source"/> collection.`</typeparam>
+    /// <param name="source">The collection to enumerate.</param>
+    /// <param name="predicate">A function to test each element for a condition.</param>
+    /// <returns><see langword="true"/> if there is only one element in the collection; <see langword="false"/> otherwise.</returns>
+    /// <remarks>
+    /// This should be equivalent to calling <see cref="Enumerable.SingleOrDefault{TSource}(IEnumerable{TSource}, Func{TSource, bool})"/>
+    /// combined with a catch that returns <see langword="null"/>.
+    /// </remarks>
+    public static TSource? SingleWhenOnly<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
```

Comment:
I said I was good with the name, but I changed my mind. Sorry about that. Now that I see the code, this feels like `SingleOrDefaultWhenOnly`, except it doesn't really (because we expect a `.Single()` to throw when there are multiple.  

A `.FirstOrDefault()` would look something like

```csharp
foreach(TSource element in source)
{
  if (predicate(element) return element;
}
```

But this is a first or default when there are multiple, so more like Single than First. This extension method needs to have really good documentation around it

Reply:
from chat:

Decided to go with `DefaultIfNotSingle()`. It doesn't have quite the same discoverability in IntelliSense because it doesn't start with "single", but the name is a lot clearer.

Reply:
Fixed!